### DISCUSSION TITLE
How to track boxing Long.valueOf calls?

### DISCUSSION THREAD

**Discussion by zuevmaxim** [async-profiler/async-profiler#1020] [Q&A] [answered]
I have the following code example that uses a lot of boxing calls
```
public class BoxingExample {
    public static void main(String[] args) {
        Long last = null;
        HashSet<Long> longs = new HashSet<>();
        for (long i = 0; i < Long.MAX_VALUE; i++) {
            Long boxed = box(i);
            longs.add(boxed);
            longs.remove(last);
            last = boxed;
        }
        System.out.println(longs.size());
    }

    private static Long box(long i) {
        return Long.valueOf(i);
    }
}
```
I expect them to be present in the profile, but I see just a few of those, but then they look to be inlined. I'm interested in whether this is expected?
<img width="798" alt="Screenshot 2024-10-11 at 11 03 15" src="https://github.com/user-attachments/assets/3df27811-0ae2-46d5-bb2e-d5b53b643291">
<img width="1025" alt="Screenshot 2024-10-11 at 11 06 11" src="https://github.com/user-attachments/assets/3702acac-5664-4d5f-b93b-602514282710">

**Comment by apangin** [comment, marked as answer]
Thank you for your quesion.

Normally, you would see a hot allocating methog on a flame graph, even if it is inlined.
However, `Long.valueOf` is not a regular method: it is a JVM [intrinsic](https://github.com/openjdk/jdk/blob/309b929147e7dddfa27879ff31b1eaad271def85/src/java.base/share/classes/java/lang/Long.java#L994), HotSpot treats it specially.
When JIT compiler encounters a call to intrinsified boxing method, it does not compile target method, but replaces the call with a special code sequence without constructing a frame, therefore, you'll see a caller method on a flame graph.

By adding `--lines` argument to the jfr converter, you'll get more detailed flame graph, where the hottest frame will point to the line where `Long.valueOf` is called:
<img width="299" alt="lines" src="https://github.com/user-attachments/assets/1b6d38c4-03c3-4a6e-a9ff-33f9ad563df8">
<img width="272" alt="valueOf" src="https://github.com/user-attachments/assets/2dcc6e5e-1174-47b2-867e-ec6e7484904f">

The best way to detect boxing is to collect an allocation profile.
Sampled allocations go through the VM runtime; at the points where non-leaf runtime calls are made, Java stack is fully consistent with the source code, i.e., you'll see `Long.valueOf` frames as expected.
<img width="615" alt="alloc" src="https://github.com/user-attachments/assets/c7e404ae-5c39-4416-9234-e74db0ee4c21">

**Reply by zuevmaxim** [reply]
Thank you!
